### PR TITLE
DROOLS-6181 : Allow sorting in guided decision table when clicking the column name

### PR DESCRIPTION
**Pull request by Rikkola**
**JIRA**: [DROOLS-6181](https://issues.redhat.com/browse/DROOLS-6181)

**Referenced Pull Requests**:
* https://github.com/kiegroup/kie-wb-common/pull/3593
* https://github.com/kiegroup/drools-wb/pull/1474

[Link for the war](https://www.dropbox.com/s/xy2zuhxz96fngsi/business-central-7.52.0-SNAPSHOT-wildfly19.war?dl=0)

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

### CODE REVIEW

**Comment by Rikkola** [comment]
I'll up the coverage and fix the code smelss.

**Comment by manstis** [comment]
Nice to see it added @Rikkola (and sorry, it was me that never carried it over from 5.x to 6.x).

**Comment by Rikkola** [comment]
Broken tests look like I broke them, fixing.

**Comment by Rikkola** [comment]
@jomarko @karreiro The tests should pass now ( at least they did locally ).

**Review by karreiro — APPROVED** [review]
Thank you, @Rikkola! LGTM and works quite well :-)

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/GuidedDecisionTableUiModelTest.java
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/GuidedDecisionTableUiModelTest.java:48]: please update the test method name to provide more information what is being tested

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/synchronizers/impl/RowSynchronizerTest.java
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/synchronizers/impl/RowSynchronizerTest.java:513]: seems we  need to apply formatting

**Review by jomarko — COMMENTED** [review]
### Acceptance criteria
@Rikkola please notice the implementation is in conflict with acceptance criteria. We implemented double click to start sorting. However jira says single click. We need to fix this inconsistency. Please update the jira as I do nto think you are going to reimplement the trigger mechanism.

### Javadoc
Neither this PR nor kie-wb-common PR contains info what the heavily used argument `final List<Integer> rowOrder` means. Could we please document it at least in one interface?

### Tests
Thank you for the added coverage. I would like to do some changes. Maybe more will come in future. Not sure if will be able automate more complex scenarios. https://github.com/Rikkola/drools-wb/pull/16

### Manual review
Starting right now

**Comment by jomarko** [comment]
### 01
Sorting merged view doesn't work https://issues.redhat.com/browse/DROOLS-6220

**Comment by jomarko** [comment]
### 02
Screenshot from 2021-03-25 22-59-27

Clicking selected cell causing sorting, but I do not know the algorithm used in such case for sorting. Could you please clarify?

**Comment by jomarko** [comment]
### 03
Sorting doesn't work properly in graph https://issues.redhat.com/browse/DROOLS-6221

**Comment by jomarko** [comment]
### 04
Sorting of local dates doesn't work - it sorts like text '01-Apr-2021' is less than '01-Mar-2021'. However I think no worth to report. t is blocked by https://issues.redhat.com/browse/DROOLS-4461. What was reported two years back without fix landing in. What is your opinion?


I am done with review. It is your turn now @Rikkola . Thank you.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java:1015]: Actual sorting is done in uiModel with the help of `TableSortComparator`. That then returns the the new row order with each row number in a list. I chose to do the sort in UIModel since we sort by visible values.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java:49]: Consumer is better and cleaner.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/synchronizers/impl/ModelSynchronizerImpl.java
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/synchronizers/impl/ModelSynchronizerImpl.java:418]: To be honest I followed this style since it was used in other methods. The benefit of this seems to be that the code below is not called.

```
public void updateSystemControlledColumnValues() {
        systemControlledColumnValuesSynchronizer.updateSystemControlledColumnValues();
}
```

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java:1015]: thank you

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java:49]: thank you

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GridRowComparatorTest.java
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GridRowComparatorTest.java:66]: I need to use the `baseGridData` to set the values, since direct access to the rows is not allowed.

**Comment by jomarko** [comment]
@Rikkola FYI
```
[INFO] There is 1 error reported by Checkstyle 8.19 with /home/jomarko/redhat/github/kiegroup/drools-wb/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/target/checkstyle-rules.xml ruleset.
[ERROR] src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java:[22,8] (imports) UnusedImports: Unused import - com.google.gwt.core.client.Callback.
```

**Comment by jomarko** [comment]
@Rikkola FYI
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project drools-wb-guided-dtable-editor-client: Compilation failure: Compilation failure: 
[ERROR] /home/jomarko/redhat/github/kiegroup/drools-wb/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandlerTest.java:[69,75] incompatible types: com.google.gwt.core.client.Callback cannot be converted to java.util.function.Consumer<org.uberfire.ext.wires.core.grids.client.model.GridColumn>
[ERROR] /home/jomarko/redhat/github/kiegroup/drools-wb/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandlerTest.java:[82,76] incompatible types: com.google.gwt.core.client.Callback cannot be converted to java.util.function.Consumer<org.uberfire.ext.wires.core.grids.client.model.GridColumn>
```

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GridRowComparatorTest.java
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GridRowComparatorTest.java:66]: ok, thank you

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java:59]: I am sorry for the inconvenience. As discussed privately. This needs to be removed. Otherwise cells can not be edited by mouse.

**Review by jomarko — COMMENTED** [review]
Thank you for updates, however I was wrong with one proposal. Apologies.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java:59]: No worries

Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableSortGridWidgetMouseEventHandler.java]: Removed

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/synchronizers/impl/ModelSynchronizerImpl.java
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/synchronizers/impl/ModelSynchronizerImpl.java:412]: ok, thank you, keep it as it is

**Comment by jomarko** [comment]
### 04
Reported as https://issues.redhat.com/browse/DROOLS-6224

**Review by jomarko — APPROVED** [review]
Thank you for updates @Rikkola 

As last check I started also internal selenium test suite and no regression were found.

**Comment by Rikkola** [comment]
```
dtable-editor-client: Compilation failure: Compilation failure: 
13:22:16  [ERROR] /home/jenkins/workspace/KIE/master/pullrequest/drools-wb-master.pr/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/TableSortComparatorTest.java:[111,83] incompatible types: capture#1 of ? cannot be converted to int
13:22:16  [ERROR] /home/jenkins/workspace/KIE/master/pullrequest/drools-wb-master.pr/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/TableSortComparatorTest.java:[156,83] incompatible types: capture#2 of ? cannot be converted to int
13:22:16  [ERROR] -> [Help 1]
```
Not getting this locally.

 jenkins retest this

**Comment by jomarko** [comment]
@Rikkola also `GuidedDecisionTableSortGridWidgetMouseEventHandlerTest#testTargetIsNotAColumn` can not pass. We do nto return false from `GuidedDecisionTableSortGridWidgetMouseEventHandler#handleHeaderCell`

We should either:
- remove test
- update `handleHeaderCell` to return false in case `gridColumn` is `nulll`

**Comment by Rikkola** [comment]
@jomarko Thanks, I think best to remove it since I only originally added it to catch the failure that basically would never happen.

**Comment by jomarko** [comment]
@Rikkola jenkins faisl with the same. I can not reproduce locally, however I guess the issue is we cast to (int). Probably we need cast to (Integer)

lets try replace:
```java
extracting(row -> (int) row.getCells().get(1).getValue().getValue())
```

with
```java
extracting(row -> (Integer) row.getCells().get(1).getValue().getValue())
```

**Comment by Rikkola** [comment]
@jomarko Thanks, need to try that. It all works for me locally.

**Review by jomarko — APPROVED** [review]
Approving. Once green is build, we can merge. 

However If more code changes (in src/main) needed, then ask me for a re-review please.

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers can I get a merge please.

**Comment by Rikkola** [comment]
jenkins do fdb

Let's run this just in case.

**Pull request merged**
